r/src/arrow_exports.h

```diff
@@ -22,6 +22,7 @@
 #include "./arrow_cpp11.h"
 
 #if defined(ARROW_R_WITH_ARROW)
+#include <arrow/dataset/api.h>
```

Comment:
Why do you need this (big) header here now?

Reply:
This was because of these classes that were not available in the fw header: 

```cpp
R6_HANDLE(arrow::dataset::DirectoryPartitioning, "DirectoryPartitioning")
R6_HANDLE(arrow::dataset::HivePartitioning, "HivePartitioning")
```

forward declaring them seems to do the trick.

Reply:
👍 We should probably add the forward declarations to arrow/dataset/type_fwd.h